app/models/queries/filters/shared/custom_fields/hierarchy.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Queries
+  module Filters
+    module Shared
+      module CustomFields
+        class Hierarchy < Base
+          def ar_object_filter?
+            true
+          end
+
+          def value_objects
+            CustomField::Hierarchy::Item
+              .where(id: @values)
+              .map { |item| HierarchyItemFilterAdapter.new(item:) }
+          end
+        end
+
+        class HierarchyItemFilterAdapter
+          attr_reader :name
+
+          delegate :id, to: :item
+
+          def initialize(item:)
+            @item = item
+            @name = item.label
+          end
+
+          private
+
+          attr_accessor :item
+        end
```

Comment:
@ulferts it's because of the expected values on the [query_filter_instance_representer.rb](https://github.com/opf/openproject/blob/dev/lib/api/v3/queries/filters/query_filter_instance_representer.rb#L76-L100). It seems to be designed to operate on a type of object and we needed the adapter so that it would render correctly.

Same for the path we added to the `path_helper.rb`